### PR TITLE
feat: support keylog and pcap mode in gnutls

### PR DESCRIPTION
**Pull request by yuweizzz**
Support keylog and pcap mode in gnutls
Supported version: gnutls 3.7.9/3.8.1/3.8.3
Release: Debian 12/ubuntu 23.10/ubuntu 24.04

### CODE REVIEW

File: kern/gnutls_masterkey.h [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+#define GNUTLS_TLS1_2 4
+#define GNUTLS_TLS1_3 5
+#define GNUTLS_DTLS1_0 201
+#define GNUTLS_DTLS1_2 202
+
+// Supported version: gnutls 3.7.9/3.8.1/3.8.3
+// Release: Debian 12/ubuntu 23.10/ubuntu 24.04
+
+// gnutls_session_int->security_parameters
+#define GNUTLS_SESSION_INT_SECURITY_PARAMETERS 0x0
```
Comment (cfc4n) [inline kern/gnutls_masterkey.h:119]: Do all versions of GNUTLS have the same offset address?

Do we need to determine the version of the class library and select the corresponding bytecode file, just like openssl?

**Review by cfc4n — CHANGES_REQUESTED** [review]

File: kern/gnutls_masterkey.h (hunk 1 above)
Comment (yuweizzz) [reply kern/gnutls_masterkey.h:119]: I not plan to support too much versions of GNUTLS, just the default version which installed in popular Linux distribution, this three version use same offset.

File: cli/cmd/gnutls.go [hunk 2]
```diff
@@ -35,17 +37,26 @@ ecapture gnutls
 ecapture gnutls --hex --pid=3423
 ecapture gnutls -l save.log --pid=3423
 ecapture gnutls --gnutls=/lib/x86_64-linux-gnu/libgnutls.so
+ecapture gnutls -m keylog -k ecapture_gnutls_key.og
```
Comment (cfc4n) [inline cli/cmd/gnutls.go:40]: Why use the file suffix `og`?

**Review by cfc4n — COMMENTED** [review]
The results tested on ubuntu 22.04 are not as expected.

```shell
sudo bin/ecapture gnutls -m keylog
2024-11-03T09:45:18Z INF AppName="eCapture(旁观者)"
2024-11-03T09:45:18Z INF HomePage=https://ecapture.cc
2024-11-03T09:45:18Z INF Repository=https://github.com/gojue/ecapture
2024-11-03T09:45:18Z INF Author="CFC4N <cfc4ncs@gmail.com>"
2024-11-03T09:45:18Z INF Description="Capturing SSL/TLS plaintext without a CA certificate using eBPF. Supported on Linux/Android kernels for amd64/arm64."
2024-11-03T09:45:18Z INF Version=linux_arm64:0.8.9-20241031-a995110:5.15.0-121-generic
2024-11-03T09:45:18Z INF Listen=localhost:28256
2024-11-03T09:45:18Z INF eCapture running logs logger=
2024-11-03T09:45:18Z INF the file handler that receives the captured event eventCollector=
2024-11-03T09:45:18Z INF listen=localhost:28256
2024-11-03T09:45:18Z INF https server starting...You can update the configuration file via the HTTP interface.
2024-11-03T09:45:18Z WRN ========== module starting. ==========
2024-11-03T09:45:18Z INF Kernel Info=5.15.163 Pid=10195
2024-11-03T09:45:18Z INF BTF bytecode mode: CORE. btfMode=0
2024-11-03T09:45:18Z INF GnuTlsProbe init eBPFProgramType=KeyLog model=KeyLog
2024-11-03T09:45:18Z INF module initialization. isReload=false moduleName=EBPFProbeGNUTLS
2024-11-03T09:45:18Z INF Module.Run()
2024-11-03T09:45:18Z INF BPF bytecode loaded bytecode filename=user/bytecode/gnutls_kern_core.o
2024-11-03T09:45:18Z INF gnutls binary path binaryPath=/lib/aarch64-linux-gnu/libgnutls.so.30 elfType=2
2024-11-03T09:45:18Z INF target all process.
2024-11-03T09:45:18Z INF perfEventReader created mapSize(MB)=4
2024-11-03T09:45:18Z INF module started successfully. isReload=false moduleName=EBPFProbeGNUTLS
2024-11-03T09:45:21Z INF CLIENT_RANDOM save success ClientRandom=f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a TlsVersion=GNUTLS_TLS1_3
^C2024-11-03T09:45:26Z INF Module closed,message recived from Context
2024-11-03T09:45:26Z INF iModule module close
2024-11-03T09:45:26Z INF bye bye.

sudo cat ecapture_gnutls_key.og
CLIENT_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
SERVER_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
EXPORTER_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a ce902f553705e6e13413a70329d6c089585fe6cf01031a9c0000000000000000
CLIENT_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 4ae8778b8b84365f4a72cce42f3bfd534a7f3ec1c3cc95b50000000000000000
SERVER_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 039d90842422683273582d0048013e87b5b830771ba127cb0000000000000000
```

The results of `CLIENT_HANDSHAKE_TRAFFIC_SECRET` and `SERVER_HANDSHAKE_TRAFFIC_SECRET` are missing.

File: cli/cmd/gnutls.go (hunk 2 above)
Comment (yuweizzz) [reply cli/cmd/gnutls.go:40]: follow the value of openssl part: https://github.com/gojue/ecapture/blob/master/cli/cmd/tls.go#L53

**Comment by yuweizzz** [comment]
> The results tested on ubuntu 22.04 are not as expected.
> 
> #### Terminal 1
> ```shell
> sudo bin/ecapture gnutls -m keylog
> 2024-11-03T09:45:18Z INF AppName="eCapture(旁观者)"
> 2024-11-03T09:45:18Z INF HomePage=https://ecapture.cc
> 2024-11-03T09:45:18Z INF Repository=https://github.com/gojue/ecapture
> 2024-11-03T09:45:18Z INF Author="CFC4N <cfc4ncs@gmail.com>"
> 2024-11-03T09:45:18Z INF Description="Capturing SSL/TLS plaintext without a CA certificate using eBPF. Supported on Linux/Android kernels for amd64/arm64."
> 2024-11-03T09:45:18Z INF Version=linux_arm64:0.8.9-20241031-a995110:5.15.0-121-generic
> 2024-11-03T09:45:18Z INF Listen=localhost:28256
> 2024-11-03T09:45:18Z INF eCapture running logs logger=
> 2024-11-03T09:45:18Z INF the file handler that receives the captured event eventCollector=
> 2024-11-03T09:45:18Z INF listen=localhost:28256
> 2024-11-03T09:45:18Z INF https server starting...You can update the configuration file via the HTTP interface.
> 2024-11-03T09:45:18Z WRN ========== module starting. ==========
> 2024-11-03T09:45:18Z INF Kernel Info=5.15.163 Pid=10195
> 2024-11-03T09:45:18Z INF BTF bytecode mode: CORE. btfMode=0
> 2024-11-03T09:45:18Z INF GnuTlsProbe init eBPFProgramType=KeyLog model=KeyLog
> 2024-11-03T09:45:18Z INF module initialization. isReload=false moduleName=EBPFProbeGNUTLS
> 2024-11-03T09:45:18Z INF Module.Run()
> 2024-11-03T09:45:18Z INF BPF bytecode loaded bytecode filename=user/bytecode/gnutls_kern_core.o
> 2024-11-03T09:45:18Z INF gnutls binary path binaryPath=/lib/aarch64-linux-gnu/libgnutls.so.30 elfType=2
> 2024-11-03T09:45:18Z INF target all process.
> 2024-11-03T09:45:18Z INF perfEventReader created mapSize(MB)=4
> 2024-11-03T09:45:18Z INF module started successfully. isReload=false moduleName=EBPFProbeGNUTLS
> 2024-11-03T09:45:21Z INF CLIENT_RANDOM save success ClientRandom=f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a TlsVersion=GNUTLS_TLS1_3
> ^C2024-11-03T09:45:26Z INF Module closed,message recived from Context
> 2024-11-03T09:45:26Z INF iModule module close
> 2024-11-03T09:45:26Z INF bye bye.
> ```
> 
> ### Terminal 2
> Trigger network behavior. On ubuntu 22.04, `git` uses the `libguntls` class library. `git pull`
> 
> ### Terminal 3
> View results
> 
> ```shell
> sudo cat ecapture_gnutls_key.og
> CLIENT_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
> SERVER_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
> EXPORTER_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a ce902f553705e6e13413a70329d6c089585fe6cf01031a9c0000000000000000
> CLIENT_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 4ae8778b8b84365f4a72cce42f3bfd534a7f3ec1c3cc95b50000000000000000
> SERVER_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 039d90842422683273582d0048013e87b5b830771ba127cb0000000000000000
> ```
> 
> The results of `CLIENT_HANDSHAKE_TRAFFIC_SECRET` and `SERVER_HANDSHAKE_TRAFFIC_SECRET` are missing.
> 
> Also, the contents of several other keys don't look right, and the last few bytes shouldn't all be `0`.
> 
> It seems that the `offset` address is not universal.

yes, the gnutls version installed of ubuntu 22.04 is 3.7.3, this offset doesn't support it.

**Comment by cfc4n** [comment]
I think it is necessary to automatically identify the version number and select the corresponding offset.

You can read strings from the `.rodata` section of `so file`'s ELF and search for `Enabled GnuTLS ([0-9\.]+) logging...` to determine the version number.

https://github.com/gnutls/gnutls/blob/aee590db7e60345dd8aa3127bc7fcb428ecde9fd/lib/global.c#L267

image

**Comment by yuweizzz** [comment]
> I think it is necessary to automatically identify the version number and select the corresponding offset. At least versions 3.6 and later must be supported.
> 
> You can read strings from the `.rodata` section of `so file`'s ELF and search for `Enabled GnuTLS ([0-9\.]+) logging...` to determine the version number.
> 
> https://github.com/gnutls/gnutls/blob/aee590db7e60345dd8aa3127bc7fcb428ecde9fd/lib/global.c#L267
> 
> image

I will check this part.

**Review by cfc4n — CHANGES_REQUESTED** [review]
Missed the offset address generator?

**Comment by yuweizzz** [comment]
Current only the C file; I will patch it later.

File: cli/cmd/gnutls.go (hunk 2 above)
Comment (xxxxxliil) [reply cli/cmd/gnutls.go]: https://github.com/gojue/ecapture/pull/436/files#diff-e8e66257f925faa6708802f4512557ebaab2558cd841f914c7f83897421adc48

这可能是一个在将近一年前引入的 typo。因为我尝试关键词 `.og file` 搜索时看到的大多数内容都与 `.ogg` 格式相关，而且根据相关上下文推测应该是 `.log` 而不是 `.og`。如果选一个合适的后缀的话我会尝试 `.keylog.txt`。

Comment (yuweizzz) [reply cli/cmd/gnutls.go]: `.log` 应该足够了。

File: utils/gnutls_offset.sh [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      continue
+    fi
+    echo "git checkout ${tag}"
+    git checkout ${tag}
+    echo "Generating ${header_file}"
+
+    # init
+    ./bootstrap
+    ./configure
+    clang -I gnulib/lib/ -I lib/ -I . offset.c -o offset
```
Comment (cfc4n) [inline utils/gnutls_offset.sh:47]: `lib/` is not the correct path, the correct path should be `lib/includes`?

At least it reported an error in my development environment (ubuntu 22.04 arm64).

```shell
In file included from offset.c:8:
./lib/gnutls_int.h:81:10: fatal error: 'gnutls/gnutls.h' file not found
#include <gnutls/gnutls.h>
         ^~~~~~~~~~~~~~~~~
1 error generated.
```

File: utils/gnutls_offset.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    if [[ -f ${header_file} ]]; then
+      echo "Skip ${header_file}"
+      continue
+    fi
+    echo "git checkout ${tag}"
+    git checkout ${tag}
+    echo "Generating ${header_file}"
+
+    # init
```
Comment (cfc4n) [inline utils/gnutls_offset.sh:44]: During initialization, some class libraries are missing and need to be installed in advance.

```shell
sudo apt install gnulib libtool gtk-doc-tools nettle-dev libtasn1-6-dev libunistring-dev libev-dev libunbound-dev
```

File: utils/gnutls_offset.sh [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    ./bootstrap
```
Comment (cfc4n) [inline utils/gnutls_offset.sh:45]: `./bootstrap --skip-po --force --no-bootstrap-sync`  

Add some parameters to ignore unnecessary behavior and other anomalies.

File: utils/gnutls_offset.sh [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    ./configure
```
Comment (cfc4n) [inline utils/gnutls_offset.sh:46]: Disable **p11-kit** to reduce library dependencies during build.
` ./configure --without-p11-kit`

File: cli/cmd/gnutls.go [hunk 7]
```diff
@@ -35,17 +37,27 @@ ecapture gnutls
 ecapture gnutls --hex --pid=3423
 ecapture gnutls -l save.log --pid=3423
 ecapture gnutls --gnutls=/lib/x86_64-linux-gnu/libgnutls.so
+ecapture gnutls -m keylog -k ecapture_gnutls_key.og --ssl_version=3.7.9
```
Comment (cfc4n) [inline cli/cmd/gnutls.go:40]: Please change `ecapture_gnutls_key.og` to `ecapture_gnutls_key.log`. The previous writing was a spelling mistake.

**Review by cfc4n — APPROVED** [review]

File: utils/gnutls_offset.sh (hunk 4 above)
Comment (cfc4n) [reply utils/gnutls_offset.sh:44]: Around line 43, some warnings are output? If an error is reported during execution, please ensure that the class library is installed successfully, or refer to [gnutls github](https://github.com/gnutls/gnutls)

File: utils/gnutls_offset.sh (hunk 3 above)
Comment (yuweizzz) [reply utils/gnutls_offset.sh]: Maybe my local env is not pure, I will try it in new container.

File: utils/gnutls_offset.sh (hunk 4 above)
Comment (yuweizzz) [reply utils/gnutls_offset.sh:44]: `gnulib` is no needed, because the gnutls will use it as a submodule.

File: cli/cmd/gnutls.go (hunk 7 above)
Comment (yuweizzz) [reply cli/cmd/gnutls.go:40]: the part is finished in https://github.com/gojue/ecapture/pull/654/commits/f514903293dfc1d1afe245a766b02bf323b8b8d5

File: utils/gnutls_offset.sh [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    # clean up
+    make clean
+  done
+
+  rm offset.c
+}
+
+# install deps
+apt install -y \
```
Comment (cfc4n) [inline utils/gnutls_offset.sh:64]: Is `sudo` missing?

```shell
bash utils/gnutls_offset.sh
check file exists: /home/cfc4n/project/ecapture_654/deps/gnutls/.git
check directory exists: /home/cfc4n/project/ecapture_654/deps/gnutls
E: Could not open lock file /var/lib/dpkg/lock-frontend - open (13: Permission denied)
E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), are you root?
```

**Review by cfc4n — CHANGES_REQUESTED** [review]

**Comment by cfc4n** [comment]
The captured key is sometimes empty, and sometimes the last few bytes are 0, which seems to be a bug.

For testing methods, see: https://github.com/gojue/ecapture/pull/654#pullrequestreview-2411816586

```shell
sudo cat ecapture_gnutls_key.og
CLIENT_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
SERVER_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
EXPORTER_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a ce902f553705e6e13413a70329d6c089585fe6cf01031a9c0000000000000000
CLIENT_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 4ae8778b8b84365f4a72cce42f3bfd534a7f3ec1c3cc95b50000000000000000
SERVER_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 039d90842422683273582d0048013e87b5b830771ba127cb0000000000000000
```

File: utils/gnutls_offset.sh (hunk 8 above)
Comment (yuweizzz) [reply utils/gnutls_offset.sh:64]: I use `root` to install package,  fix it later.

**Comment by yuweizzz** [comment]
> The captured key is sometimes empty, and sometimes the last few bytes are 0, which seems to be a bug.
> 
> For testing methods, see: [#654 (review)](https://github.com/gojue/ecapture/pull/654#pullrequestreview-2411816586)
> 
> ```shell
> sudo cat ecapture_gnutls_key.og
> CLIENT_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
> SERVER_HANDSHAKE_TRAFFIC_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 0000000000000000000000000000000000000000000000000000000000000000
> EXPORTER_SECRET f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a ce902f553705e6e13413a70329d6c089585fe6cf01031a9c0000000000000000
> CLIENT_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 4ae8778b8b84365f4a72cce42f3bfd534a7f3ec1c3cc95b50000000000000000
> SERVER_TRAFFIC_SECRET_0 f5413ff872ec489fa15601f89461d4f5880c4c2dd1993d8fc8f03845a5a8d10a 039d90842422683273582d0048013e87b5b830771ba127cb0000000000000000
> ```

work good at amd64, need to check out what happened in arm64. Could you try it again in amd64?

File: utils/gnutls_offset.sh (hunk 8 above)
Comment (xxxxxliil) [reply utils/gnutls_offset.sh]: 要不要在安装依赖前检查发行版……？我不确定例如 aosc 这种也用 apt 的发行版是否也存在上述包，然后也要考虑下其他非 debian 且非 apt/deb 发行版的依赖安装

File: utils/gnutls_offset.sh (hunk 4 above)
Comment (cfc4n) [reply utils/gnutls_offset.sh:44]:
```shell
Submodule path 'devel/libtasn1': checked out '2b7ee166f031bea63c5e7badcedc69f1cea69c89'
./bootstrap: 69: ./bootstrap.conf: gnulib/gnulib-tool: not found
./bootstrap: Bootstrapping from checked-out gnutls sources...
./bootstrap: consider installing git-merge-changelog from gnulib
./bootstrap: getting gnulib files...
Submodule 'gnulib' (https://gitlab.com/libidn/gnulib-mirror.git) registered for path 'gnulib'
Cloning into '/home/ubuntu/ecapture/deps/gnutls/gnulib'...
```

I tested it on an amd64 Ubuntu 22.04 system, and it seems that I still need to clone this repository.

File: utils/gnutls_offset.sh (hunk 8 above)
Comment (cfc4n) [reply utils/gnutls_offset.sh]: 这个脚本是用在开发ecapture的环境的，不是用在使用ecapture场景，所以，可以只支持ubuntu。

**Comment by cfc4n** [comment]
On ubuntu 22.04 amd64 operating system, it works fine.
```shell
sudo bin/ecapture gnutls -m keylog
2024-11-11T23:36:48+08:00 INF AppName="eCapture(旁观者)"
2024-11-11T23:36:48+08:00 INF HomePage=https://ecapture.cc
2024-11-11T23:36:48+08:00 INF Repository=https://github.com/gojue/ecapture
2024-11-11T23:36:48+08:00 INF Author="CFC4N <cfc4ncs@gmail.com>"
2024-11-11T23:36:48+08:00 INF Description="Capturing SSL/TLS plaintext without a CA certificate using eBPF. Supported on Linux/Android kernels for amd64/arm64."
2024-11-11T23:36:48+08:00 INF Version=linux_amd64:0.8.9-20241110-956a95c:5.15.0-122-generic
2024-11-11T23:36:48+08:00 INF Listen=localhost:28256
2024-11-11T23:36:48+08:00 INF eCapture running logs logger=
2024-11-11T23:36:48+08:00 INF the file handler that receives the captured event eventCollector=
2024-11-11T23:36:48+08:00 WRN ========== module starting. ==========
2024-11-11T23:36:48+08:00 INF Kernel Info=5.15.163 Pid=46567
2024-11-11T23:36:48+08:00 INF BTF bytecode mode: CORE. btfMode=0
2024-11-11T23:36:48+08:00 INF GnuTlsProbe init eBPFProgramType=KeyLog model=KeyLog
2024-11-11T23:36:48+08:00 INF module initialization. isReload=false moduleName=EBPFProbeGNUTLS
2024-11-11T23:36:48+08:00 INF Module.Run()
2024-11-11T23:36:48+08:00 INF listen=localhost:28256
2024-11-11T23:36:48+08:00 INF https server starting...You can update the configuration file via the HTTP interface.
2024-11-11T23:36:48+08:00 INF GnuTLS version found Version=3.7.3
2024-11-11T23:36:48+08:00 INF GnuTLS binary path binaryPath=/lib/x86_64-linux-gnu/libgnutls.so.30 elfType=2
2024-11-11T23:36:48+08:00 INF BPF bytecode loaded bytecode filename=user/bytecode/gnutls_3_7_3_kern_core.o
2024-11-11T23:36:48+08:00 INF target all process.
2024-11-11T23:36:49+08:00 INF perfEventReader created mapSize(MB)=4
2024-11-11T23:36:49+08:00 INF module started successfully. isReload=false moduleName=EBPFProbeGNUTLS
2024-11-11T23:36:51+08:00 INF CLIENT_RANDOM save success ClientRandom=8ddeccf3e6045511b0923a3c5c17d49719ca0f95857b9f38108453f1a0a4da87 TlsVersion=GNUTLS_TLS1_3
^C2024-11-11T23:36:55+08:00 INF Module closed,message recived from Context
2024-11-11T23:36:55+08:00 INF iModule module close
2024-11-11T23:36:55+08:00 INF bye bye.

sudo cat ecapture_gnutls_key.log
CLIENT_HANDSHAKE_TRAFFIC_SECRET 8ddeccf3e6045511b0923a3c5c17d49719ca0f95857b9f38108453f1a0a4da87 bf861c22ebd3aec93315f7c5aa489485c32d6711e2e371b86511b8dd05b6cec6
SERVER_HANDSHAKE_TRAFFIC_SECRET 8ddeccf3e6045511b0923a3c5c17d49719ca0f95857b9f38108453f1a0a4da87 f02278d21c02dff1b69e0a6f2d2203d194962ace1c560d1449b512338134f638
EXPORTER_SECRET 8ddeccf3e6045511b0923a3c5c17d49719ca0f95857b9f38108453f1a0a4da87 dbcb8e86adc748c641649bb213ecdcccc2bc60f009280fa377c75ede4024456e
CLIENT_TRAFFIC_SECRET_0 8ddeccf3e6045511b0923a3c5c17d49719ca0f95857b9f38108453f1a0a4da87 13eb72395ab0920a58599f6a8d405fbb9215dfebd79d988184815167a61930a5
SERVER_TRAFFIC_SECRET_0 8ddeccf3e6045511b0923a3c5c17d49719ca0f95857b9f38108453f1a0a4da87 d97779ec0972003d1dd1f8ddabc706709c8e4d09c415d5a1d61d81099fe43d89
```

File: utils/gnutls_offset.sh [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    echo "git clone gnutls to ${GNUTLS_DIR}"
+    git clone https://github.com/gnutls/gnutls.git ${GNUTLS_DIR}
+  fi
+fi
+
+
+function run() {
+  git fetch --tags
+  cp -f ${PROJECT_ROOT_DIR}/utils/gnutls_offset.c ${GNUTLS_DIR}/offset.c
+  main_version="3.8"
```
Comment (cfc4n) [inline utils/gnutls_offset.sh:28]: Why the later `3.8` version? How to generate the offset addresses of `3.6.*` and `3.7.*`?

**Review by cfc4n — APPROVED** [review]
I tried to regenerate `gnutls_*_*_*_kern.c` on arm64, but it seems that only two can be generated. Please increase the generation of other files, thank you very much.

File: utils/gnutls_offset.sh (hunk 9 above)
Comment (cfc4n) [reply utils/gnutls_offset.sh:28]: Perhaps, it is reasonable to only support versions after `3.8.*` to generate offset addresses.

Comment (yuweizzz) [reply utils/gnutls_offset.sh:28]: need to change `main_version` and loop condition here to support others version, we can do this locally, `3.8.x` is the newest version.

**Pull request merged**